jina/types/arrays/neural_ops.py

```diff
@@ -108,7 +84,99 @@ def match(
                 if d.id in self:
                     d = Document(d, copy=True)
                     d.pop('matches')
-                _q.matches.append(d, scores={m_name: _dist}, copy=False)
+                _q.matches.append(d, scores={metric_name: _dist}, copy=False)
+
+    def _match(self, darray, cdist, limit, normalization, metric_name):
+        """
+        Computes the matches between self and `darray` loading `darray` into main memory.
+        :param darray: the other DocumentArray or DocumentArrayMemmap to match against
+        :param cdist: the distance metric
+        :param limit: the maximum number of matches, when not given
+                      all Documents in `another` are considered as matches
+        :param normalization: a tuple [a, b] to be used with min-max normalization,
+                                the min distance will be rescaled to `a`, the max distance will be rescaled to `b`
+                                all values will be rescaled into range `[a, b]`.
+        :param metric_name: if provided, then match result will be marked with this string.
+        :return: distances and indices
+        """
+        is_sparse = False
+
+        if isinstance(darray[0].embedding, np.ndarray):
+            x_mat = np.stack(self.get_attributes('embedding'))
+            y_mat = np.stack(darray.get_attributes('embedding'))
+        else:
+            import scipy.sparse as sp
+
+            if sp.issparse(darray[0].embedding):
+                x_mat = sp.vstack(self.get_attributes('embedding'))
+                y_mat = sp.vstack(darray.get_attributes('embedding'))
+                is_sparse = True
+
+        if is_sparse:
+            dists = cdist(x_mat, y_mat, metric_name, is_sparse=is_sparse)
+        else:
+            dists = cdist(x_mat, y_mat, metric_name)
+
+        dist, idx = top_k(dists, min(limit, len(darray)), descending=False)
```

Comment:
this is not properly considering that `limit` can be `None` according to the `typehint`